### ISSUE TITLE
commit-log

### ISSUE THREAD

**Issue by asbjornenge** [asbjornenge/hyper-symposium#1]
## Topic

I want to explore building a distributed commit-log using node, and I'm looking for a design discussion.
## Why

The point of this commit-log is to gather all events in distributed systems in an immutable fashion. This has all kinds of nice properties like snapshots, timetravel, debugging etc. Another cool thing would be if we can easily build dynamic indexes on top of this structure, providing a "view" into subsets of the log. These indexes or views could act as backends for applications. And so we can throw out the traditional mutable database.

Other systems like this include [kafka](http://kafka.apache.org/) and [samza](https://samza.apache.org/).
## Requirements

My requirements are:
- Distributed (obviously)
- Something I can snapshot
- Something I can build dynamic views (indexes) on top of
## Questions
- Does such a thing already exists? 
- Is it a good idea to make one with node? 
- What properties should it have? 
- Internal logical structure (what type of hash tree) ?
- External/"physical" structure (how do we [store]() the data on disk) ?

**PS!** _Distributed_ is a word we throw around a lot and it's meaning sometimes get a bit lost, so I would like to clarify; I don't require the commit-log to be distributed for the sake of being distributed (like p2p systems), but for the properties it provides, like high-availability and scalability.

Imho we are already far along this path with modules like [hyperlog](https://www.npmjs.com/package/hyperlog), [hyperemitter](https://www.npmjs.com/package/hyperemitter) and [hyperfs](https://www.npmjs.com/package/hyperfs).  Big question is how do we put them together to make this thing :stuck_out_tongue_winking_eye: 

//cc @mafintosh @mcollina

**Comment by mcollina** [comment]
I wrote HyperEmitter to back something like the thing you are describing.

Here are my takes:
1. it should be eventually consistent
2. there is an event log, with application specific views on top
3. it should support snapshots
4. it should allow to remove history from the event log, and use a snapshot to restore

(the 4th point is needed to reduce the time needed to sync a new peer, while building dynamic indexes)

Regarding your questions, I don't have many ideas, but I will put some thoughts on them.

A different solution to this problem is [Raft](https://raftconsensus.github.io/), and we have a node implementation [skiff](https://www.npmjs.com/package/skiff). However, it has strong consensus. 

Adding in @mcdonnelldean and @pgte.

**Comment by mcdonnelldean** [comment]
What is the use case? It seems like you are describing an distributed event store, kind of like NEventStore but with distributed bits.

I don't want to add thoughts until I know more :D

**Comment by asbjornenge** [comment]
@mcollina 

> I wrote HyperEmitter to back something like the thing you are describing.

I had a feeling & was kinda hoping you did :stuck_out_tongue_closed_eyes: I agree with your points. 

With regards to point 4; I'm not sure about removing data from the log. I see it as an immutable data-structure, where removing data would be like removing commits from git...? New peers of the log would (hopefully) only sync partials of a tree (depending on how we distribute), but I can live with that not being super fast (I think). Building the indexes/views however, needs to be super fast. To support that we need to enable fast replay of events. I think this is something we should support at an API level, so that different type of views can be built on top of our log. Perhaps we can do some magic with snapshot + sync here?

I actually have a few different distributed systems where I want to use this. One is built on top of [serf](https://serfdom.io/) where the event propagation is already handled. One where I'm actually using HyperEmitter (lots of iot stuff around my house actually, heh), and another where I'm breaking up a monolith bit by bit and building out microservices. For that last one I'm using [faye](http://faye.jcoglan.com/) to propagate events.

So, for me at least, it makes sense to build this **separate from the event system**. 

**I think the bit we need to figure out is how we structure the data, both logically and "physically" to allow massive scale (by distribution) and fast "replay of events".**

Maybe we just need a distributed storage layer backing hyperlog? Are there any modules for distributing leveldb? I found [this](http://hackingdistributed.com/2013/06/17/hyperleveldb/) but I have way to little experience with leveldb to say anything about it. Would love @mafintosh input here :wink: 

Another thing I'm pondering is if it makes sense to have 1 huge log for the entire system, or if it makes more sense to have 1 log for each "topic" (I want to load everything together into some big data query engine, but I could live with merging them prior to that as a separate step).

@mcdonnelldean 

One of the main reasons I've created this repo is to avoid duplicate effort. I hadn't seen NEventStore before, so thanks for sharing :smile: 

My main use-case is that I want to store every single event happening in my distributed system to capture it's state. I want to take snapshots of my entire distributed application and use that for backup, debugging, easily create new environments etc. 

Also; in addition, if I could build this log structure in such a way that I could construct what I refer to as views on top of it, that would remove much of my requirements to manage mutable databases to drive applications. A _view_ can be an index or a key-val store or anything that drive applications. I think of it as database-views for the log structure.

YIKES, that's a long reply :stuck_out_tongue_winking_eye: Hope at least some of it made sense. :hear_no_evil: :see_no_evil: :speak_no_evil: 

tldr; HALP!

**Comment by asbjornenge** [comment]
I like many of the ideas in [Datomic](http://www.datomic.com/). If we could provide a **layer of facts** (or facilitate the creation of such a thing) that should really speed up building views. Some interesting node work [here](https://www.npmjs.com/package/level-immutable). The [datom](http://pelle.github.io/Datomic/2012/07/08/thinking-in-datomic/) is also very interesting as a way of structuring these facts.

**Comment by asbjornenge** [comment]
Had an idea :bulb: 

What about, instead of trying to build a log **_service**_, we think of the log more as "**_infrastructure**_"; every service has one. So we end up with this huge collection of **_micrologs**_ instead of this one big log service.

_What are the tradeoffs? How would we implement?_

Rebuilds could be fast since we now have 1000 services contributing. It does however make all our services stateful in a sense, since they all now hold a piece of the puzzle. This log layer could however be decoupled from the service logic and only the log needs to remain.

Sorry if this is a bit brainstormy, I just needed to unload :stuck_out_tongue_closed_eyes: Maybe move to a separate issue?

Anyway, thoughts? :sparkles:
